File: packages/format/src/av1/leb128.ts
import { FormatError } from "../errors.js";

export interface Av1Leb128 {
  readonly value: number;
  readonly length: number;
}

/** Read a canonical unsigned LEB128 value bounded to safe integers. */
export function readAv1Leb128(
  bytes: Uint8Array,
  offset: number,
  path = "av1.leb128"
): Readonly<Av1Leb128> {
  if (!(bytes instanceof Uint8Array) || !Number.isSafeInteger(offset) || offset < 0) {
    throw new FormatError("PROFILE_INVALID", "AV1 LEB128 input is invalid", { path });
  }
  let value = 0n;
  let length = 0;
  for (; length < 8; length += 1) {
    const byte = bytes[offset + length];
    if (byte === undefined) {
      throw new FormatError("PROFILE_INVALID", "AV1 LEB128 is truncated", {
        path,
        offset: offset + length
      });
    }
    value |= BigInt(byte & 0x7f) << BigInt(length * 7);
    if ((byte & 0x80) === 0) {
      const byteLength = length + 1;
      if (byteLength > 1 && value < (1n << BigInt((byteLength - 1) * 7))) {
        throw new FormatError("PROFILE_INVALID", "AV1 LEB128 is non-canonical", {
          path,
          offset
        });
      }
      if (value > BigInt(Number.MAX_SAFE_INTEGER)) {
        throw new FormatError("PROFILE_INVALID", "AV1 LEB128 is unsafe", {
          path,
          offset
        });
      }
      return Object.freeze({ value: Number(value), length: byteLength });
    }
  }
  throw new FormatError("PROFILE_INVALID", "AV1 LEB128 exceeds eight bytes", {
    path,
    offset
  });
}

export function encodeAv1Leb128(value: number): Uint8Array {
  if (!Number.isSafeInteger(value) || value < 0) {
    throw new FormatError("PROFILE_INVALID", "AV1 LEB128 value is invalid");
  }
  const bytes: number[] = [];
  let remaining = BigInt(value);
  do {
    let byte = Number(remaining & 0x7fn);
    remaining >>= 7n;
    if (remaining !== 0n) byte |= 0x80;
    bytes.push(byte);
  } while (remaining !== 0n);
  return Uint8Array.from(bytes);
}

File: flutter/packages/aval_format/lib/src/av1/leb128.dart
/// Canonical unsigned LEB128 reading/writing bounded to safe integers.
///
/// Dart port of `packages/format/src/av1/leb128.ts`.
library;

import 'dart:typed_data';

import '../checked_integer.dart' show maxSafeInteger;
import '../errors.dart';

class Av1Leb128 {
  const Av1Leb128({required this.value, required this.length});

  final int value;
  final int length;

  @override
  bool operator ==(Object other) =>
      other is Av1Leb128 && other.value == value && other.length == length;

  @override
  int get hashCode => Object.hash(value, length);
}

/// Read a canonical unsigned LEB128 value bounded to safe integers.
Av1Leb128 readAv1Leb128(Uint8List bytes, int offset,
    [String path = 'av1.leb128']) {
  if (offset < 0 || offset > maxSafeInteger) {
    throw FormatError(
      FormatErrorCode.profileInvalid,
      'AV1 LEB128 input is invalid',
      FormatErrorDetails(path: path),
    );
  }
  var value = 0;
  var length = 0;
  for (; length < 8; length += 1) {
    if (offset + length >= bytes.length) {
      throw FormatError(
        FormatErrorCode.profileInvalid,
        'AV1 LEB128 is truncated',
        FormatErrorDetails(path: path, offset: offset + length),
      );
    }
    final byte = bytes[offset + length];
    value |= (byte & 0x7f) << (length * 7);
    if ((byte & 0x80) == 0) {
      final byteLength = length + 1;
      if (byteLength > 1 && value < (1 << ((byteLength - 1) * 7))) {
        throw FormatError(
          FormatErrorCode.profileInvalid,
          'AV1 LEB128 is non-canonical',
          FormatErrorDetails(path: path, offset: offset),
        );
      }
      if (value > maxSafeInteger) {
        throw FormatError(
          FormatErrorCode.profileInvalid,
          'AV1 LEB128 is unsafe',
          FormatErrorDetails(path: path, offset: offset),
        );
      }
      return Av1Leb128(value: value, length: byteLength);
    }
  }
  throw FormatError(
    FormatErrorCode.profileInvalid,
    'AV1 LEB128 exceeds eight bytes',
    FormatErrorDetails(path: path, offset: offset),
  );
}

Uint8List encodeAv1Leb128(int value) {
  if (value < 0 || value > maxSafeInteger) {
    throw FormatError(
      FormatErrorCode.profileInvalid,
      'AV1 LEB128 value is invalid',
    );
  }
  final bytes = <int>[];
  var remaining = value;
  do {
    var byte = remaining & 0x7f;
    remaining >>= 7;
    if (remaining != 0) byte |= 0x80;
    bytes.add(byte);
  } while (remaining != 0);
  return Uint8List.fromList(bytes);
}
